Add end-to-end tests for Volume Group Snapshot

manishym opened the pull request:
This commit introduces comprehensive e2e tests for the Volume Group Snapshot feature to ensure robust validation and functionality. Addresses issue #1080.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
End to end test of group snapshot feature.
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes [#1080 ](https://github.com/kubernetes-csi/external-snapshotter/issues/1080) External snapshotter

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
NONE
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

manishym commented:
@xing-yang can you please take a look? https://github.com/kubernetes/kubernetes/pull/126326

carlory commented:
/ok-to-test

xing-yang commented:
Can you add this full url in the PR description? https://github.com/kubernetes-csi/external-snapshotter/issues/1080.  Otherwise it automatically points to an issue opened in k/k.

xing-yang commented:
/kind feature

xing-yang commented:
/release-note-none

manishym commented:
/test pull-kubernetes-linter-hints

manishym commented:
/retest

manishym committed:
Add CRDs for VolumeGroupSnapshots management

This commit introduces three new CustomResourceDefinitions (CRDs) for managing volume group snapshots within Kubernetes clusters:
1. VolumeGroupSnapshotClasses: Defines the properties and parameters required for volume group snapshot classes.
2. VolumeGroupSnapshotContents: Details the structure and management of on-disk group snapshot contents.
3. VolumeGroupSnapshots: Specifies user requests and properties for creating or binding to group snapshots.

Each CRD is equipped with comprehensive specs, including fields like deletionPolicy, driver, and creationTime, tailored to enhance management capabilities and integration with the CSI driver specifications.

Approved API references:
- VolumeGroupSnapshotClasses: kubernetes-csi/external-snapshotter#814
- VolumeGroupSnapshotContents and VolumeGroupSnapshots: kubernetes-csi/external-snapshotter#1068

Signed-off-by: Manish <myathnal@redhat.com>

manishym committed:
Enable volume group snapshots in CSI hostpath plugin

This commit updates the CSI hostpath plugin configuration by adding the `--enable-volume-group-snapshots=true` argument to the container startup parameters. This change allows the CSI hostpath plugin to support volume group snapshots functionality, enhancing testing capabilities for features that require group snapshot support in end-to-end tests.

Signed-off-by: Manish <myathnal@redhat.com>

manishym commented:
[ginkgo_snapshot_test.log](https://github.com/user-attachments/files/16657754/ginkgo_snapshot_test.log)

Logs from my testing.

manishym commented:
I ran the tests in `pull-kubernetes-e2e-gce-storage-snapshot ` also in my setup, they are passing. The failure we are seeing here must be some infrastructure issue. Here are snapshot test logs from my local setup

[snapshot_test.log](https://github.com/user-attachments/files/16659302/snapshot_test.log)

manishym commented:
/retest

manishym committed:
Add end-to-end tests for Volume Group Snapshot

This commit introduces comprehensive e2e tests for the
Volume Group Snapshot feature to ensure robust validation and functionality.
Addresses issue [kubernetes#1080 ](kubernetes-csi/external-snapshotter#1080).

Signed-off-by: Manish <myathnal@redhat.com>

Madhu-1 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,124 @@
+	return snapshot
+}
+
+// SnapshotResource represents a snapshot class, a snapshot and its bound snapshot contents for a specific test case
```
```suggestion
// GroupSnapshotResource represents a volume group snapshot class, a group snapshot, and its bound group snapshot contents for a specific test case
```

Madhu-1 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,124 @@
+	Vgsclass   *unstructured.Unstructured
+}
+
+// CreateSnapshot creates a VolumeSnapshotClass with given SnapshotDeletionPolicy and a VolumeSnapshot
```
function comment need to be updated to reflect `CreateGroupSnapshot`

Madhu-1 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,124 @@
+
+	groupSnapshot, err = dc.Resource(utils.VolumeGroupSnapshotGVR).Namespace(groupSnapshot.GetNamespace()).Create(ctx, groupSnapshot, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+	ginkgo.By("Waiting for snapshot to be ready")
```
waiting for group snapshot to be ready?

Madhu-1 commented on test/e2e/storage/testsuites/group_snapshottable.go:
```diff
@@ -0,0 +1,226 @@
+	dInfo := driver.GetDriverInfo()
+	ok := false
+	_, ok = driver.(storageframework.VoulmeGroupSnapshottableTestDriver)
+	if !dInfo.Capabilities[storageframework.CapSnapshotDataSource] || !ok {
```
`CapSnapshotDataSource` is not opt for group snapshot we need to add new capabilities as drivers can support snapshot but not groupsnapshot

Madhu-1 commented on test/e2e/storage/testsuites/group_snapshottable.go:
```diff
@@ -0,0 +1,226 @@
+	ok := false
+	_, ok = driver.(storageframework.VoulmeGroupSnapshottableTestDriver)
+	if !dInfo.Capabilities[storageframework.CapSnapshotDataSource] || !ok {
+		e2eskipper.Skipf("Driver %q does not support snapshots - skipping", dInfo.Name)
```
```suggestion
e2eskipper.Skipf("Driver %q does not support group snapshots - skipping", dInfo.Name)
```

Madhu-1 commented on test/e2e/storage/testsuites/group_snapshottable.go:
```diff
@@ -0,0 +1,226 @@
+			}
+
+			cleanup := func(ctx context.Context) {
+				// for _, snapshot := range groupTest.snapshots {
```
Remove commented code

Madhu-1 commented on test/e2e/storage/utils/group_snapshot.go:
```diff
@@ -0,0 +1,60 @@
+			},
+			"driver":         snapshotter,
+			"parameters":     parameters,
+			"deletionPolicy": "Delete",
```
policy should be configurable or it will be always Delete?

yati1998 replied:
+1, I guess it should be configurable.

xing-yang replied:
We don't want to leave storage resources on the system after finishing the tests. So if you are using "Retain", make sure to change it to "Delete" before deleting the group snapshot.

Madhu-1 commented on hack/local-up-cluster.sh:
```diff
@@ -1071,7 +1071,9 @@ function start_csi_snapshotter {
         ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/volumesnapshots/volume-snapshot-controller/rbac-volume-snapshot-controller.yaml"
         ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/volumesnapshots/volume-snapshot-controller/volume-snapshot-controller-deployment.yaml"
-
+        ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/volume-group-snapshots/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml"
```
i think we need to change the order as below
* Create CRD's
* Create RBAC
* Create deployment

Madhu-1 commented on hack/local-up-cluster.sh:
```diff
@@ -1071,7 +1071,9 @@ function start_csi_snapshotter {
-
+        ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/volume-group-snapshots/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml"
+        ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/volume-group-snapshots/groupsnapshot.storage.k8s.io_volumegroupsnapshotclasses.yaml"
+        ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" apply -f "${KUBE_ROOT}/cluster/addons/volume-group-snapshots/groupsnapshot.storage.k8s.io_volumegroupsnapshotcontents.yaml"
```
i think we also need to update RBAC for volumegroupsnapshot operation isn't it?

xing-yang commented on cluster/addons/volume-group-snapshots/groupsnapshot.storage.k8s.io_volumegroupsnapshotclasses.yaml:
```diff
@@ -0,0 +1,94 @@
+---
```
It turns out we can't install VolumeGroupSnapshots CRDs in cluster addons because they are for cloud provider jobs. 
See the following.  Here's how we enabled VolumeSnapshot testing in a GCE cluster. However, GCE CSI driver does not support VolumeGroupSnapshot yet, so the job can't run there.
https://github.com/kubernetes/test-infra/blob/6b9389ee5a64a70556b31b4bf7d7557a831f1fc9/config/jobs/kubernetes/sig-storage/sig-storage-gce-config.yaml#L106

xing-yang commented:
Take a look at this PR which added a new job in the Kind cluster: https://github.com/kubernetes/test-infra/pull/31717
We can use this job and add VolumeGroupSnapshot feature there.

We can't use cluster addon to install CRDs and snapshot-controller, etc, but we'll still need to install them as part of this job.

manishym commented:
/retest

manishym commented:
/retest

yati1998 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,124 @@
+	Vgsclass   *unstructured.Unstructured
+}
+
+// CreateGroupSnapshot creates a VolumeSnapshotClass with given SnapshotDeletionPolicy and a VolumeSnapshot
```
please update the function description to include volumegroupsnapshotclass and volumegroupsnapshot.

yati1998 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,124 @@
+	return gsclass, groupSnapshot
+}
+
+func (r *GroupSnapshotResource) CleanupResource(ctx context.Context, timeouts *framework.TimeoutContext) error {
```
good to have a comment to mention the cleanup resource is volumegroupsnapshotclass.

manishym commented:
/retest

xing-yang commented on cluster/addons/volumesnapshots/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshotclasses.yaml:
```diff
@@ -0,0 +1,94 @@
+---
```
Can you move these from cluster/addons to under testing-manifests?

xing-yang commented on test/e2e/testing-manifests/storage-csi/external-snapshotter/volume-group-snapshots/run_group_snapshot_e2e.sh:
```diff
@@ -0,0 +1,327 @@
+}
+
+# up a cluster with kind
+create_cluster() {
```
Do you need to install a cluster manually here?

manishym replied:
Yes

xing-yang commented on test/e2e/storage/testsuites/group_snapshottable.go:
```diff
@@ -0,0 +1,221 @@
+
+			}
+
+			ginkgo.It("should create snapshots for multiple volumes in a pod", func(ctx context.Context) {
```
Can you add a test for static provisioning?

gnufied commented:
/assign

manishym commented:
/retest

manishym commented:
/retest

Madhu-1 commented on test/e2e/storage/framework/testpattern.go:
```diff
@@ -60,6 +60,8 @@ var (
 	// PreprovisionedCreatedSnapshot represents a snapshot type for pre-provisioned snapshot
 	PreprovisionedCreatedSnapshot TestSnapshotType = "PreprovisionedSnapshot"
+
+	GroupVolumeSnapshot TestSnapshotType = "GroupVolumeSnapshot"
```
`GroupVolumeSnapshot` to `VolumeGroupSnapshot`

Madhu-1 commented on test/e2e/storage/framework/testpattern.go:
```diff
@@ -318,6 +320,14 @@ var (
 	}
+
+	// GroupVolumeSnapshotDelete is TestPattern for "Groupsnapshot"
+	GroupVolumeSnapshotDelete = TestPattern{
```
GroupVolumeSnapshotDelete to VolumeGroupSnapshotDelete

Madhu-1 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,126 @@
+	return gsclass, groupSnapshot
+}
+
+// CleanupResource deletes the VolumeGroupSnapshotClass and VolumeGroupSnapshot objects using a dynamic client.
```
we are deleting only the class not the volumeGroupSnapshot, this comment need to be updated as well.

Madhu-1 commented on test/e2e/storage/testsuites/group_snapshottable.go:
```diff
@@ -0,0 +1,221 @@
+type volumeGroupSnapshottableTest struct {
+	config      *storageframework.PerTestConfig
+	pods        []*v1.Pod
+	volumeGroup [3][]*storageframework.VolumeResource
```
why we have 3 limit here?

manishym replied:
Its not a limit, it is the number of snapshots we are creating.

Madhu-1 commented on test/e2e/storage/framework/group_snapshot_resource.go:
```diff
@@ -0,0 +1,126 @@
+					"selector": map[string]interface{}{
+						"matchLabels": map[string]interface{}{
+							"group": groupName,
+						},
```
IMO, we need to make both key and value as input for this function.

Madhu-1 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -79,10 +80,14 @@ var CSISuites = append(BaseSuites,
 	InitVolumePerformanceTestSuite,
 	InitReadWriteOncePodTestSuite,
 	InitVolumeModifyTestSuite,
+	InitSnapshottableStressTestSuite,
+	InitVolumePerformanceTestSuite,
+	InitReadWriteOncePodTestSuite,
```
Do we need this one?

manishym replied:
No, removed

Madhu-1 commented on test/e2e/storage/utils/group_snapshot.go:
```diff
@@ -0,0 +1,64 @@
+
+var (
+
+	// VolumeGroupSnapshotGVR is GroupVersionResource for groupvolumesnapshots
```
`groupvolumesnapshots` to `volumegroupsnapshots`

Madhu-1 commented on test/e2e/storage/utils/group_snapshot.go:
```diff
@@ -0,0 +1,64 @@
+
+	// VolumeGroupSnapshotGVR is GroupVersionResource for groupvolumesnapshots
+	VolumeGroupSnapshotGVR = schema.GroupVersionResource{Group: VolumeGroupSnapshotAPIGroup, Version: "v1alpha1", Resource: "volumegroupsnapshots"}
+	// VolumeGroupSnapshotClassGVR is GroupVersionResource for groupvolumesnapshotclasses
```
`groupvolumesnapshotclasses` to `volumegroupsnapshotclasses`

manishym replied:
Fixed

gnufied commented on test/e2e/testing-manifests/storage-csi/external-snapshotter/volume-group-snapshots/run_group_snapshot_e2e.sh:
```diff
@@ -0,0 +1,329 @@
+  # NOTE: must match the number of workers above
+  NUM_NODES=2
+  # actually create the cluster
+  # TODO(BenTheElder): settle on verbosity for this script
```
Do Are we missing following steps:

```
 kind build node-image 
 kind create cluster --image kindest/node:latest 
``` 

? I am unable to run this locally otherwise.

manishym replied:
It is here: https://github.com/kubernetes/kubernetes/blob/fb7b7dd93ab0a598dae246c178dd6fbe77b2b9ff/test/e2e/testing-manifests/storage-csi/external-snapshotter/volume-group-snapshots/run_group_snapshot_e2e.sh#L74
build is called before create cluster.

gnufied replied:
hmm. By default the tests didn't work for me, unless I created image and tagged it manualy.I guess one of the problems is, by default when we build image it has tag - `localhost/kindest/node:latest` but test actually uses - `docker.io/kindest/node:latest` and it fails. I was running this with kind that uses podman underneath. I think results will be different if you were using docker may be.

gnufied commented on test/e2e/testing-manifests/storage-csi/external-snapshotter/volume-group-snapshots/run_group_snapshot_e2e.sh:
```diff
@@ -0,0 +1,329 @@
+  FOCUS="${FOCUS:-}"
+  LABEL_FILTER="${LABEL_FILTER:-}"
+  if [ -z "${FOCUS}" ] && [ -z "${LABEL_FILTER}" ]; then
+    FOCUS="\\[Conformance\\]"
```
Why would we want to run entire `Conformance` tests from here by default?

manishym replied:
This will run only if FOCUS is not set, if you see the PR here:
https://github.com/kubernetes/test-infra/commit/b4478000c126f49a4a8a5a42c4cc598fa7eac383
Focus is being set to volume group snapshot.

gnufied replied:
Understood. But why not default to that Focus here? This shell script is anyways only used for volumegroup snapshot tests.

manishym replied:
Initially my plan was to keep changes to this script minimum and patch it in the yaml in test_infra, but that is error prone, so I added the entire script. I would prefer to keep the changes minimum, if in case the script in https://kind.sigs.k8s.io/dl/latest/linux-amd64.tgz changes and we need to update the changes.

gnufied commented on test/e2e/storage/framework/volume_group_snapshot_resource.go:
```diff
@@ -0,0 +1,126 @@
+	volumeGroupSnapshot := getVolumeGroupSnapshot(groupName, pvcNamespace, gsclass.GetName())
+
+	volumeGroupSnapshot, err = dc.Resource(utils.VolumeGroupSnapshotGVR).Namespace(volumeGroupSnapshot.GetNamespace()).Create(ctx, volumeGroupSnapshot, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
```
Can you please an add explanation for all `framework.ExpectNoError` ?

manishym replied:
Done

carlory commented on test/e2e/feature/feature.go:
```diff
@@ -392,6 +392,7 @@ var (
 	// TODO: document the feature (owning SIG, when to use this feature for a test)
 	VolumeSnapshotDataSource = framework.WithFeature(framework.ValidFeatures.Add("VolumeSnapshotDataSource"))
 
+	VolumeGroupSnapshotDataSource = framework.WithFeature(framework.ValidFeatures.Add("volumegroupsnapshot"))
```
please add a blank line after L395. Add a comment to indicate this feature is owned by sig-storage like L382

manishym replied:
Done

carlory commented:
which job will run these new tests?

xing-yang commented:
@carlory This is the job: https://github.com/kubernetes/test-infra/pull/33417

gnufied commented:
/test ci-kubernetes-e2e-storage-kind-volume-group-snapshots

xing-yang commented:
Hi @manishym, you need to submit another PR to test-infra to add a presumit job like this "pull-kubernetes-e2e-storage-kind-alpha-features" in this PR: https://github.com/kubernetes/test-infra/pull/31717/files.  The one you added earlier is a periodic job.

gnufied commented:
/lgtm

xing-yang commented:
/test pull-kubernetes-e2e-storage-kind-volume-group-snapshots

xing-yang commented:
/retest

xing-yang commented:
@manishym Please take a look at the job failure.

manishym commented:
/retest

xing-yang commented:
/test pull-kubernetes-e2e-storage-kind-volume-group-snapshots

manishym commented:
/retest

xing-yang commented:
/lgtm
/approve
